text/0000-never-patterns.md

```diff
@@ -0,0 +1,250 @@
+[guide-level-explanation]: #guide-level-explanation
+
+Patterns can be used on a partially uninitialized place; the basic rule is that a pattern only accesses data that is directly mentioned in the pattern: an `Ok(..)` pattern requires a discriminant read, a binding requires a full read, a wildcard `_` accesses nothing etc. 
+
+For uninhabited types (types with no valid values), normal patterns cannot be used to indicate an access since there is no data to be accessed. For this purpose, you can use the special `!` pattern.
+
+The `!` pattern is allowed on any uninhabited type (such as `enum Void {}` or `(Result<!, !>, u32)`) and does an access to the underlying data. Since there can be no such data (on pain of UB), the corresponding arm is known to be unreachable and does not need a body:
+```rust
+enum Void {}
+unsafe fn empty_tup<T>(tup: *const (T, Void)) -> ! {
+    unsafe {
+        match *tup { ! }
+    }
+}
+```
+
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+We add `!` to the syntax of patterns. A `!` pattern is accepted for any type that is visibly uninhabited (i.e. uninhabited taking into account private fields and `#[non_exhaustive]` annotations) from the item containing the match expression.
+
+A pattern that contains a `!` is called a _never pattern_. For or-patterns, all alternatives must be never patterns for the whole pattern to be too. For example `Ok(x) | Err(!)` is not a never pattern but `(_, !)` and `Ok((_, !)) | Err(!)` are.
+
+A never pattern is accepted as an arm in a match expression, and that arm takes no body nor guards.
+
+```rust
+let x: *const Option<(u32, (u32, !))> = ...;
+match *x {
+    None => { ... }
+    Some((_, !)),
+}
+```
+
+A never pattern is also accepted in other places a pattern would be (`if let`, destructuring `let`, etc.), and makes code that corresponds to that branch unreachable.
+
+```rust
+enum Void {}
+impl Void {
+    fn ex_falso<T>(!: Self) -> T {}
+}
+```
+
+In terms of both semantics and exhaustiveness checking, a `!` pattern behaves like a binding. E.g. a `Some(!),` arm (absent match ergonomics) is semantically equivalent to `Some(x) => match x {}`. Indeed, they both indicate a load at a place of an uninhabited type, followed by an unreachable body.
+
+Never patterns interact with match ergonomics as expected: a `!` pattern is allowed on e.g. `&(T, !)`.
+
+A never pattern may be linted as "unreachable" if there is no well-behaved execution that can reach it. For example, if it was guaranteed that reading the discriminant of `Result<T, !>` could never return `Err`, the `Err(!)` pattern would be linted as unreachable. Never patterns are not otherwise linted as unreachable even when removing them would be accepted.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+The main drawback is that this new pattern is unlike other patterns (e.g. it doesn't require an arm body) which adds some maintenance burden. It's otherwise a simple feature.
```

Comment:
how bad would it be to this RFC if the "feature" of skipping the arm body be removed :shrug: save some major changes to the syntax that messes with macros and `syn` etc.

just type-check that the arm body always has type `const !` to replace that "a never pattern is always unreachable" error. 

```rust
match x {
    // note: using `const { unreachable!() }` instead of only `unreachable!()`
    // to tell the compile actually enforce that the arm is never invoked
    Some(!) => const { unreachable!() },
    None => "ok",
}
```

tbh it is very strange that *only* the `match` syntax has this omission feature. you can't do it in an `if let` for instance. given how `#![feature(never_patterns)]` works today you need to write:

```rust
if let Some(!) = x {
    // you still need a body here! otherwise you get "E0308 mismatched types"!
    const { unreachable!() }
} else {
    "ok"
}
```

(i don't think the "empty block" proposal in https://github.com/rust-lang/rfcs/pull/3719#issuecomment-2440138142 works because an empty block does not indicate the block is actually unreachable. the `!` expression is cute but it think it may conflict with unary-not `!(false)`.)

Reply:
The other syntax that has an equivalent is or-patterns: `Ok(x) | Err(!)` is a valid pattern because the never case "doesn't count" for the purpose of ensuring that all alternatives have the same binding. Otherwise you're right, match arms are special.

Reply:
If we don't want the "bodiless arm" syntax, I'd propose that we allow this:
```rust
match x {
    Some(!) => {}
    None => "ok",
}
```
this would consider the whole arm body to diverge for the purposes of type-checking, similar to how we typecheck `=> { return; () }` or `=> { panic!(); () }`. In fact I already implemented something like this for function arguments:
```rust
enum Void {}
fn foo(!: Void) -> u32 {} // typechecks because the body diverges
```

There's one case where the "bodiless arm" is particularly nice though: the empty match. Today `match *foo {}` is allowed for `foo: *const !`, which violates the intent that matching on empty types inside pointers should be explicit. My intent after this RFC is to deprecate this case and eventually force users to write `match *foo { ! }` instead. This is less nice if they have to write `match *foo { ! => {} }`.

Reply:
The specialty of `Ok(x) | Err(!)` manifests only *after* parsing which you have built the AST which then you can check for "E0408 variable is not bound in all patterns". The impact is not at the same level as the proposed syntax `match expr { pat(!) }` that changes the grammar and also the resulting AST/HIR structure.

Reply:
> The `!` expression is cute but it think it may conflict with unary-not `!(false)`.

how so? if you're trying to have a `!` expression, you can just say it can't be followed by a function call or other ambiguous tokens, if you really want to call it, you'd have to write `(!)(false)` (not that that would type check...)

Reply:
or just say `!` can't be followed by any postfix or binary expressions -- problem solved.

Reply:
@programmerjake sure we can always adjust the parser to disambiguate, but I think introducing a whole new syntax of expression with special handling (like `..`) for this rather obscure feature is not a place we should spend the weirdness budget on. And as I've shown above `const { unreachable!() }` is basically a spelling of the `!` expression that already works on stable Rust.

----

@Nadrieril 

> similar to how we typecheck `=> { return; () }` or `=> { panic!(); () }`.

Hmm it does not typecheck?

```rust
#![feature(never_type)]

fn f(a: Option<&!>) -> &str {
    match a {
        Some(_) => { return ""; () },
        // error[E0308]: mismatched types
        // expected `&str`, found `()`
        None => "ok",
    }
}
```

> `fn foo(!: Void) -> u32 {} // typechecks because the body diverges`

I suppose that means the `if let` example not working for `{}` is a bug?

```rust
#![feature(never_patterns)]
#![feature(never_type)]

fn g(a: Option<&!>) -> &str {
    if let Some(!) = a {
        // error[E0308]: mismatched types
        // expected `&str`, found `()`
    } else {
        "ok"
    }
}
```

> This is less nice if they have to write `match *foo { ! => {} }`.

I don't think the need for never-patterns is common enough to justify changing the `match` syntax just to save that `=> {}` 😅

Reply:
(`=> { return ""; () }` doesn't type check in that `match`, but `=> { return ""; }` or `=> { return ""; (); }` does. IIUC diverging blocks (currently) only have type `!` for type checking purposes if they do not have an explicit tail expression.)

Reply:
> for this rather obscure feature is not a place we should spend the weirdness budget on.

ok, since `{}` seems to type check, I think using `! => {}` as the match arm is good enough.

Reply:
> I don't think the need for never-patterns is common enough to justify changing the match syntax just to save that `=> {}` 😅

Fair. I was hoping we could allow it just for this special case but idk if that's worth it.

> I suppose that means the if let example not working for {} is a bug?

Yep

Reply:
My slight concern with a syntax like `! => {}` is that it reads to me like "types which match ! are permitted, please ignore them and return `()`" rather than "types matching ! are unreachable/UB".

I haven't thought too much about this, but would something like `! => !` type check? (if we temporarily ignore a dependency on the never type).

Reply:
@dhedey `=> !` can certainly type-check but then we need to introduce the new "`!`" expression in additional to the pattern and type.

> but prevent e.g. ! => { x = x + 1; panic!("Side effect!"); } which I would assume also has a type of !?

A non-empty block (or non-`!` value) should raise the unreachable_code lint like now:

```rust
// warning: unreachable statement
fn f(ptr: &mut u8, !: &!) {
//                 - any code following a never pattern is unreachable
    *ptr += 1;
//  ^^^^^^^^^^ unreachable statement
}
```

Reply:
If we have to have an arm, IMO `=> !` makes most sense. It does still look quite funny, though...